### PR TITLE
[AIR] Add Scaling Config validation

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds a `ScalingConfigDataClass.validate_config` classmethod to allow for a generic way of validating ScalingConfigs by allowing only certain keys.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/config.py
Comment (krfricke) [inline python/ray/ml/config.py:137]: As mentioned in the previous PR, let's remove these arguments. The enclosing method can re-raise the exception so that the affected key names are in the stack trace, but we should avoid passing strings just for raising error messages

**Review by krfricke — CHANGES_REQUESTED** [review]

File: python/ray/ml/config.py
Comment (Yard1) [reply python/ray/ml/config.py:137]: I wonder if we can return a list of bad keys instead. I think a single exception would be better from the user perspective than a long stack trace

File: python/ray/ml/config.py
Comment (gjoliver) [inline python/ray/ml/config.py:11]: please sort the imports man. otherwise it's much harder to see whether I need to import something or not.

File: python/ray/ml/config.py
Comment (gjoliver) [inline python/ray/ml/config.py:146]: I think I had this comment on the other PR as well.
why would a prohibited_key live on default_config dict?

File: python/ray/ml/config.py
Comment (matthewdeng) [inline python/ray/ml/config.py:119]: Where is this called?

Comment (Yard1) [reply python/ray/ml/config.py:119]: In a separate pr - https://github.com/ray-project/ray/pull/23850/

File: python/ray/ml/config.py
Comment (Yard1) [reply python/ray/ml/config.py:146]: We use this method to restrict the user to a subset of keys (allowed_keys). Therefore, all the keys that are not allowed are prohibited from being set by the user.

File: python/ray/ml/config.py
Comment (amogkam) [reply python/ray/ml/config.py:119]: Can we call this in the base Trainer class and also have a class attribute containing the allowed keys? Then each sub-trainer can override this class attribute. This will force all the Trainers to be explicit on which keys are allowed, and makes it easier to see this when reading the code/API docs.

File: python/ray/ml/config.py
Comment (amogkam) [inline python/ray/ml/config.py:118]: Can we make this into a regular function instead of a class method?

We are not using any class attributes anywhere, and we are planning to make ScalingConfigDataclass a user facing public API eventually, so I don't think we should expose this method for users to call themselves.

Also, the validation only makes sense in the context of Trainers for now, so I'm not sure if it should be a method on the dataclass itself.

File: python/ray/ml/config.py
Comment (amogkam) [inline python/ray/ml/config.py:141]: Should we raise ValueError instead? This error is about the input config being invalid rather than a key not existing in a dict.

File: python/ray/ml/config.py
Comment (amogkam) [inline python/ray/ml/config.py:155]:
```suggestion
                f"Key(s) {bad_keys} are not allowed for this Trainer. "
```

File: python/ray/ml/config.py
Comment (amogkam) [inline python/ray/ml/config.py:153]: This currently only checks if the value is set to something other than the default value. But the user can still set a prohibited key to the default value and expect it to be honored. This seems fine for now, but I'm wondering if there's a way to address this case as well.

**Review by amogkam — CHANGES_REQUESTED** [review]
Thanks @Yard1, this looks great so far! I think we should also do some more refactoring to make the allowed keys explicit for each Trainer

File: python/ray/ml/config.py
Comment (gjoliver) [reply python/ray/ml/config.py:119]: like the idea of defining allowed keys explicitly.

**Review by ericl — COMMENTED** [review]
+1 on @amogkam 's suggestions. I think you want this end state right?

```
class Trainer:
   _scaling_config_supported_keys = []  # Not scalable by default
   
   @classmethod
   def validate_config(scaling_config):
       ... default validator ...
 ```
 
 ```
 class XGBoostTrainer:
    _scaling_config_supported_keys = ["num_workers", "use_gpu"]
 ```
 
I'd also like to see this PR show the whole integration path with Trainer, rather than adding a utility function without showing how works with other classes. This will make it easier to review the interfaces rather than the implementation.

File: python/ray/ml/config.py
Comment (matthewdeng) [inline python/ray/ml/config.py:121]: Another idea is to have each `Trainer` define its own `ScalingConfigDataClass` subclass, so that users can tell from the public documentation exactly what's supported.

Comment (amogkam) [reply python/ray/ml/config.py:121]: Yeah we talked about this approach in the past, but I think for now we decided to just have a single ScalingConfig

```
Decision:
Start with sidestep: Trainer(
	**rllib_specific_configs, 
	scaling_config=ScalingConfig(**common_configs))
Let's revisit this in the future: RLlibScalingConfig(**rllib_specific_configs, **common_configs)
```

File: python/ray/ml/config.py
Comment (bveeramani) [inline python/ray/ml/config.py:125]: Nit: You could wrap `allowed_keys` with backticks (i.e., ``) so that it renders as code in the documentation.

File: python/ray/ml/config.py
Comment (bveeramani) [inline python/ray/ml/config.py:130]: Nit: Same with `ScalingConfigDataClass`

Comment (bveeramani) [inline python/ray/ml/config.py:130]: The docstring summary contradicts the `allowed_keys` argument description.

File: python/ray/ml/config.py
Comment (bveeramani) [reply python/ray/ml/config.py:141]: +1

File: python/ray/ml/config.py
Comment (bveeramani) [inline python/ray/ml/config.py:142]: Nit: you could use a more specific word than "valid". Like, ```Allowed key(s) {bad_allowed_keys} are not `ScalingConfigDataClass` attributes.```

File: python/ray/ml/config.py
Comment (bveeramani) [inline python/ray/ml/config.py:139]: Nit: I was kinda confused by this at first. Could improve readability to name `config.__dict__` (e.g., `scaling_config_attributes = set(default_config.__dict__)` (not sure what the best variable name would be)

File: python/ray/ml/tests/test_api.py
Comment (bveeramani) [inline python/ray/ml/tests/test_api.py:57]: Nit: it might be beneficial to split up this test into three tests.

From [Microsoft's unit testing best practices](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#avoid-multiple-acts):

> When writing your tests, try to only include one Act per test. Common approaches to using only one act include:
> * Create a separate test for each act.
> * Use parameterized tests.

> Why?
> * When the test fails it is not clear which Act is failing.
> * Ensures the test is focussed on just a single case.
> * Gives you the entire picture as to why your tests are failing.

> Multiple Acts need to be individually Asserted and it is not guaranteed that all of the Asserts will be executed. In most unit testing frameworks, once an Assert fails in a unit test, the proceeding tests are automatically considered to be failing. This can be confusing as functionality that is actually working, will be shown as failing.

File: python/ray/ml/config.py
Comment (krfricke) [reply python/ray/ml/config.py]: I would be open to raise a custom exception here that stores the keys as an attribute.

`ray.air.exceptions.ConfigError` or so

File: python/ray/ml/config.py
Comment (krfricke) [reply python/ray/ml/config.py:139]: Renamed to `invalid_keys`

File: python/ray/ml/config.py
Comment (krfricke) [reply python/ray/ml/config.py:142]: We generalized this function so resolving this comment for now

File: python/ray/ml/config.py
Comment (krfricke) [reply python/ray/ml/config.py:146]: Added two comments to make this clearer

File: python/ray/ml/config.py
Comment (krfricke) [reply python/ray/ml/config.py]: Actually we don't need this check at all. This is not a user-facing API and the invalid keys are static per Trainer. If we need this check it should be done somewhere else.

**Review by krfricke — COMMENTED** [review]
I've updated the PR to address the comments from the last feedback round:

1. The validation method has been split up in a dict and dataclass part
2. It has been moved to a utility function and generalized to arbitrary dataclass and dict objects
3. The Trainer now has a private method to return a validate scaling config dataclass
4. The code is being used in the DP/GBDP trainer classes.

I'll also update the sklearn trainer to utilize this. PTAL

File: python/ray/ml/trainer.py
Comment (Yard1) [inline python/ray/ml/trainer.py:140]:
```suggestion
    _scaling_config_allowed_keys: List[str]
```
We can actually define it like that, which would cause an exception to be raised on init if it's not overridden in a subclass

**Review by Yard1 — COMMENTED** [review]
Looks great, thanks @krfricke!

File: python/ray/ml/trainer.py
Comment (krfricke) [reply python/ray/ml/trainer.py:140]: If we do this, we should also consider turning `Trainer` into an abstract base class. Any reason why we didn't do this, yet? cc @matthewdeng @amogkam

Comment (amogkam) [reply python/ray/ml/trainer.py:140]: It currently is right?

File: python/ray/ml/train/data_parallel_trainer.py
Comment (Yard1) [reply python/ray/ml/train/data_parallel_trainer.py:263]: I think it's fine, we already call it in Trainer as well (in `as_trainable`). This is more of a way to get the dataclass inside the training loop

File: python/ray/ml/trainer.py
Comment (krfricke) [reply python/ray/ml/trainer.py:140]: Then let's do it @Yard1

Comment (krfricke) [reply python/ray/ml/trainer.py]: Ah, quick question, I've just tried:

```
import abc
from typing import List


class Base(abc.ABC):
    _wurst: List[str]


Base()
```

which works fine, so this isn't actually validated from the base class but it fails on attribute access. If that's the case, let's go back to `[]` again, as we don't want an attribute access error as the default case.

Comment (Yard1) [reply python/ray/ml/trainer.py]: Thanks for checking, must have been misremembering the behavior. Let me revert

**Review by amogkam — APPROVED** [review]
LGTM! Thanks for making the changes!

**Comment by amogkam** [comment]
Needs an approval from @krfricke before we can merge.

**Review by krfricke — APPROVED** [review]

**Review by matthewdeng — APPROVED** [review]

**Pull request merged**
